drivers/greth.c

```diff
@@ -0,0 +1,564 @@
+	}
+	mutexUnlock(state->tx_lock);
+
+	return nf ? ERR_OK : ERR_BUF;
+}
+
+
+static inline uint8_t greth_getByte(uint32_t v, int i)
+{
+	return (v >> (i * 8)) & 0xFF;
+}
+
+
+static void greth_readCardMac(greth_state_t *state)
+{
+	uint8_t *mac;
+	mac = (void *)&state->netif->hwaddr;
+
+	/* MAC LSB reg*/
+	mac[5] = greth_getByte(state->mmio->MAC_LSB, 0);
+	mac[4] = greth_getByte(state->mmio->MAC_LSB, 1);
+	mac[3] = greth_getByte(state->mmio->MAC_LSB, 2);
+	mac[2] = greth_getByte(state->mmio->MAC_LSB, 3);
+
+	/* MAC MSB reg*/
+	mac[1] = greth_getByte(state->mmio->MAC_MSB, 0);
+	mac[0] = greth_getByte(state->mmio->MAC_MSB, 1);
+}
+
+
+static void greth_showCardId(greth_state_t *state)
+{
+	uint8_t *mac;
+	mac = (void *)&state->netif->hwaddr;
+	greth_printf(state, "initizalized, MAC=%02x:%02x:%02x:%02x:%02x:%02x", mac[0], mac[1], mac[2], mac[3], mac[4], mac[5]);
+}
+
+
+static void greth_startGreth(greth_state_t *state)
+{
+	/* clear status */
+	state->mmio->STAT =
+			GRETH_STAT_IA | GRETH_STAT_TS |
+			GRETH_STAT_TA | GRETH_STAT_RA |
+			GRETH_STAT_TI | GRETH_STAT_RI |
+			GRETH_STAT_TE | GRETH_STAT_RE;
+
+	/* interrupt mask */
+	state->mmio->CTRL |= GRETH_CTRL_TI | GRETH_CTRL_RI;
+}
+
+
+static int greth_createResources(greth_state_t *state)
+{
+	if (mutexCreate(&state->irq_lock) < 0) {
+		return -1;
+	}
+	if (condCreate(&state->irq_cond) < 0) {
+		return -1;
+	}
+	return 0;
+}
+
+
+static void greth_cleanupResources(greth_state_t *state)
+{
+	if (state->irq_lock < (handle_t)-1) {
+		resourceDestroy(state->irq_lock);
+	}
+	if (state->irq_cond < (handle_t)-1) {
+		resourceDestroy(state->irq_cond);
+	}
+}
+
+
+static int greth_initDevice(greth_state_t *state, int irq, int mdio)
+{
+	state->mmio = physmmap(state->dev_phys_addr, 0x1000);
+	if (state->mmio == MAP_FAILED) {
+		return -ENOMEM;
+	};
+
+#ifdef GRETH_DISABLE_EDCL
+	int err = greth_disable_edcl(state);
+#endif
+
+	err = greth_initRings(state);
+	if (err != 0) {
+		return err;
+	}
+	greth_debug_printf(state, "Initialized GRETH Rings");
+
+	if (err != 0) {
+		greth_printf(state, "Couldn't register interrupt handler: %s (%d)", strerror(-err), err);
+		return err;
+	}
+	greth_debug_printf(state, "Interrupt handler initialized successfully");
+
+	if (mdio != 0) {
+		err = register_mdio_bus(&greth_mdio_ops, state);
+		if (err < 0) {
+			greth_printf(state, "Can't register MDIO bus:  %s (%d)", strerror(-err), err);
+			return err;
+		}
+
+		greth_debug_printf(state, "MDIO bus %d", err);
+	}
+
+	err = greth_createResources(state);
+	if (err != 0) {
+		greth_printf(state, "Couldn't create resources: %s (%d)", strerror(-err), err);
+		greth_cleanupResources(state);
+		return err;
+	}
+
+	err = interrupt(irq, greth_irqHandler, state, state->irq_cond, NULL);
+	if (err != 0) {
+		greth_printf(state, "Couldn't register interrupt handler: %s (%d)", strerror(-err), err);
+		return err;
+	}
+
+	err = beginthread(greth_irqThread, 4, state->irq_stack, sizeof(state->irq_stack), state);
```

Comment:
Priority 4 is too low for irq thread, I'd set it to 1
```suggestion
	err = beginthread(greth_irqThread, 1, state->irq_stack, sizeof(state->irq_stack), state);
```

Reply:
ok, changed